test/libyul/yulOptimizerTests/memoryStoreRemover/complex2.yul

```diff
@@ -0,0 +1,32 @@
+{
+    f()
+    let z := 100
+    pop(mload(z))
+
+    if calldataload(0) {
+        // does not read from location [20, 52)
+        pop(keccak256(z, z))
+    }
+
+    function f() {
+        let x := 10
+        let y := 20
+        // can be removed
```

Comment:
Where does the code track that we do not read from 10 after the call?

Reply:
Specifically, it's https://github.com/ethereum/solidity/pull/11192/files#diff-b652a95e7b2e5dff3527ed41ee64ffef604d00314a96050f30a282bffbddfd22R315

We store all operations that read from memory. In this case, it's just `keccak(z, z)`. We then try to prove each of the read operation can read from `[10, 42)`. If none of them reads from `[10, 42)`, we can mark `x` for removal.

Reply:
Ok and this works because all offsets in the whole file are static, right? So the optimizer does not track which function calls which and whether or not a function reverts and so on, right? So if `f` would be defined as
```
function f(x, y) { mstore(x, y) }
```
it would not remove the mstore, even if `f` is never called or if `f` is called as `f(10, 20)`. Is that correct?

Reply:
> it would not remove the mstore, even if f is never called or if f is called as f(10, 20). Is that correct?

Yes. That is correct.